internal/controllers/core/togglebutton/reconciler.go

```diff
@@ -0,0 +1,291 @@
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/tilt-dev/tilt/internal/controllers/indexer"
+	"github.com/tilt-dev/tilt/pkg/apis/core/v1alpha1"
+)
+
+const (
+	actionUIInputName = "action"
+	turnOnInputValue  = "on"
+	turnOffInputValue = "off"
+)
+
+type Reconciler struct {
+	ctrlClient            ctrlclient.Client
+	indexer               *indexer.Indexer
+	queue                 workqueue.RateLimitingInterface
+	mu                    sync.Mutex
+	lastClickProcessTimes map[string]time.Time
+}
+
+func (r *Reconciler) CreateBuilder(mgr ctrl.Manager) (*builder.Builder, error) {
+	b := ctrl.NewControllerManagedBy(mgr).
+		For(&v1alpha1.ToggleButton{}).
+		Watches(&source.Kind{Type: &v1alpha1.ConfigMap{}},
+			handler.EnqueueRequestsFromMapFunc(r.indexer.Enqueue)).
+		Owns(&v1alpha1.UIButton{})
+
+	return b, nil
+}
+
+func NewReconciler(ctrlClient ctrlclient.Client, scheme *runtime.Scheme) *Reconciler {
+	return &Reconciler{
+		ctrlClient:            ctrlClient,
+		indexer:               indexer.NewIndexer(scheme, indexToggleButton),
+		queue:                 workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "togglebutton"),
+		lastClickProcessTimes: make(map[string]time.Time),
+	}
+}
+
+func indexToggleButton(obj client.Object) []indexer.Key {
+	var result []indexer.Key
+	toggleButton := obj.(*v1alpha1.ToggleButton)
+	bGVK := v1alpha1.SchemeGroupVersion.WithKind("ConfigMap")
+
+	if toggleButton != nil {
+		if toggleButton.Spec.StateSource.ConfigMap != nil {
+			result = append(result, indexer.Key{
+				Name: types.NamespacedName{Name: toggleButton.Spec.StateSource.ConfigMap.Name},
+				GVK:  bGVK,
+			})
+		}
+	}
+	return result
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	nn := request.NamespacedName
+
+	var tb v1alpha1.ToggleButton
+	err := r.ctrlClient.Get(ctx, nn, &tb)
+	r.indexer.OnReconcile(nn, &tb)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return ctrl.Result{}, err
+	}
+
+	if apierrors.IsNotFound(err) || !tb.ObjectMeta.DeletionTimestamp.IsZero() {
+		err := r.managedOwnedUIButton(ctx, nn, nil)
+		return ctrl.Result{}, err
+	}
+
+	err = r.processClick(ctx, tb)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	err = r.manageToggleButton(ctx, &tb)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	err = r.managedOwnedUIButton(ctx, nn, &tb)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, err
+}
+
+func (r *Reconciler) processClick(ctx context.Context, tb v1alpha1.ToggleButton) error {
+	uiButton := v1alpha1.UIButton{}
+	err := r.ctrlClient.Get(ctx, types.NamespacedName{Name: uibuttonName(tb.Name)}, &uiButton)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
```

Comment:
usually you'd use the client.IgnoreNotFound idiom, as demonstrated here:
https://book.kubebuilder.io/cronjob-tutorial/controller-implementation.html#1-load-the-cronjob-by-name